### ISSUE TITLE
missing type info when use `assert isinstance/issubclass` with a generic variable

### ISSUE THREAD

**Issue by GreyElaina** [microsoft/pyright#4372]
**Describe the bug**
when use `assert isinstance/issubclass` to narrow type, this feature will hit the `TypeVar`'s `bound` instead of its actual inference.

**To Reproduce**
```py
from typing import TYPE_CHECKING, Mapping, TypeVar

class ExampleClass:
    ...

class Example(ExampleClass):
    ...

T = TypeVar("T", bound=ExampleClass)

def reproduce(x: type[T]):
    result = x()
    if TYPE_CHECKING:
        assert isinstance(result, Mapping)  # just for example
        reveal_type(result)  # Type of "result" is "<subclass of ExampleClass and Mapping>*"
    return result

reveal_type(reproduce(Example))  # Type of "reproduce(Example)" is "<subclass of ExampleClass and Mapping>*"
```

**Expected behavior**
```py
from typing import TYPE_CHECKING, Mapping, TypeVar

class ExampleClass:
    ...

class Example(ExampleClass):
    ...

T = TypeVar("T", bound=ExampleClass)

def reproduce(x: type[T]):
    result = x()
    if TYPE_CHECKING:
        assert isinstance(result, Mapping)  # just for example
        reveal_type(result)  # Type of "result" is "<subclass of type[T] and Mapping>*"
    return result

reveal_type(reproduce(Example))  # Type of "reproduce(Example)" is "<subclass of Example and Mapping>*"
```

**Screenshots or Code**

**VS Code extension or command-line**
with `Pylance`, however it's excepted that it happened with command line `pyright`

**Additional context**

**Comment by erictraut** [comment]
I'm not sure what bug you're reporting here. I don't see any problem in the code sample above.

When you use an `isinstance` type guard, the type is narrowed to the intersection of the pre-narrowed type and the type passed as the second argument to `isinstance`. The type in this case is also a type variable, so the upper bound is used, and a [conditional type](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#constrained-type-variables-and-conditional-types) is created (hence the `*` on the end of the type).

**Comment by GreyElaina** [comment]
Considering that this may be an undefined behavior, using HKT for my needs is a more generic and reasonable behavior.
